File: problems/2467_most-profitable-path-in-a-tree/solution.go
func mostProfitablePath(edges [][]int, bob int, amount []int) int {
	n := len(amount)
	adj := make([][]int, n)
	for i := range adj {
		adj[i] = make([]int, 0, 4)
	}
	for _, e := range edges {
		adj[e[0]] = append(adj[e[0]], e[1])
		adj[e[1]] = append(adj[e[1]], e[0])
	}

	parent := make([]int, n)
	depth := make([]int, n)
	seen := make([]bool, n)
	for i := range parent {
		parent[i] = -1
	}
	seen[0] = true
	order := make([]int, 0, n)
	queue := []int{0}
	for head := 0; head < len(queue); head++ {
		u := queue[head]
		order = append(order, u)
		for _, v := range adj[u] {
			if !seen[v] {
				seen[v] = true
				parent[v] = u
				depth[v] = depth[u] + 1
				queue = append(queue, v)
			}
		}
	}

	bobTime := make(map[int]int)
	t := 0
	node := bob
	for node != -1 {
		bobTime[node] = t
		t++
		node = parent[node]
	}

	income := make([]int, n)
	hasBest := false
	best := 0
	for _, u := range order {
		d := depth[u]
		bt, ok := bobTime[u]
		var gain int
		if !ok || bt > d {
			gain = amount[u]
		} else if bt == d {
			// Floor division (Python semantics) for negative amounts.
			gain = amount[u] / 2
			if amount[u] < 0 && amount[u]%2 != 0 {
				gain--
			}
		} else {
			gain = 0
		}
		if u != 0 {
			income[u] = income[parent[u]] + gain
		} else {
			income[u] = gain
		}
		if u != 0 && len(adj[u]) == 1 {
			if !hasBest || income[u] > best {
				best = income[u]
				hasBest = true
			}
		}
	}
	return best
}

File: problems/2467_most-profitable-path-in-a-tree/solution.ts
function mostProfitablePath(
    edges: number[][],
    bob: number,
    amount: number[],
): number {
    const n = amount.length;
    const adj: number[][] = Array.from({ length: n }, () => []);
    for (const [a, b] of edges) {
        adj[a].push(b);
        adj[b].push(a);
    }

    const parent = new Array<number>(n).fill(-1);
    const depth = new Array<number>(n).fill(0);
    const seen = new Array<boolean>(n).fill(false);
    seen[0] = true;
    const order: number[] = [];
    const queue: number[] = [0];
    let head = 0;
    while (head < queue.length) {
        const u = queue[head++];
        order.push(u);
        for (const v of adj[u]) {
            if (!seen[v]) {
                seen[v] = true;
                parent[v] = u;
                depth[v] = depth[u] + 1;
                queue.push(v);
            }
        }
    }

    const bobTime = new Map<number, number>();
    let t = 0;
    let node = bob;
    while (node !== -1) {
        bobTime.set(node, t);
        t++;
        node = parent[node];
    }

    const income = new Array<number>(n).fill(0);
    let best: number | null = null;
    for (const u of order) {
        const d = depth[u];
        const bt = bobTime.get(u);
        let gain: number;
        if (bt === undefined || bt > d) {
            gain = amount[u];
        } else if (bt === d) {
            gain = Math.floor(amount[u] / 2);
        } else {
            gain = 0;
        }
        income[u] = (u !== 0 ? income[parent[u]] : 0) + gain;
        if (u !== 0 && adj[u].length === 1) {
            if (best === null || income[u] > best) best = income[u];
        }
    }
    return best as number;
}
